File: src/2020/Day23.c
#include <stdio.h>

#define NCUPS 1000000
#define NMOVES 10000000
#define WRAP(n) (((n) - 1 + NCUPS) % NCUPS + 1)

typedef struct cup {
    long label;
    struct cup *next;
} cup;

int main(void) {
    static cup cups[NCUPS];

    long init[9] = {3,9,8,2,5,4,7,1,6};
    int rinit[9] = {7,3,0,5,4,8,6,2,1};

    for (long i = 0; i < NCUPS; i++)
        cups[i] = (cup) { .label = i < 9 ? init[i] : i + 1
                        , .next = cups + (i + 1) % NCUPS };

    cup *cur = cups;
    for (int i = 0; i < NMOVES; i++){
        cup *moved[3] = {cur->next, cur->next->next, cur->next->next->next};

        int dest = cur->label;
        do
            dest = WRAP(dest - 1);
        while (  dest == moved[0]->label
              || dest == moved[1]->label
              || dest == moved[2]->label );

        cur->next = moved[2]->next;
        cup *pdest = &cups[dest < 10 ? rinit[dest - 1] : dest - 1];
        moved[2]->next = pdest->next;
        pdest->next = moved[0];
        cur = cur->next;
    }

    printf("%ld\n", cups[rinit[0]].next->label
                  * cups[rinit[0]].next->next->label);

    return 0;
}

File: src/2020/Day23.hs
module Day23 where

import Control.Monad.ST
import Data.Array.ST

import AOC

complete :: [a] -> [a] -> [a]
complete (x:xs) (y:ys) = x:complete xs ys
complete    xs     []  = []
complete    []     ys  = ys

crabulate :: Int -> Int -> [Int] -> [Int]
crabulate size moves cups = runST do
    ring <- newArray_ (1, size) :: ST s (STUArray s Int Int)
    let right = readArray ring
        setRight = writeArray ring
        ringList = complete cups [1..size]
    zipWithM_ setRight ringList (tail $ cycle ringList)
    let cupsOrder 0 _ = pure []
        cupsOrder n cup = do
            next <- right cup
            (cup:) <$> cupsOrder (n - 1) next
        loop 0 _ = pure ()
        loop n cur = do
            moved@[front, _, back] <- cupsOrder 3 =<< right cur
            let dest = head $ filter (`notElem` moved) $ [cur-1,cur-2..1] ++ [size,size-1..cur+1]
            cur' <- right back
            setRight cur cur'
            setRight back =<< right dest
            setRight dest front
            loop (n - 1) cur'
    loop moves (head cups)
    cupsOrder size 1

main :: IO ()
main = do
    cups <- map digitToInt . head . lines <$> readInput
    putStrLn $ map intToDigit $ tail $ crabulate 9 100 cups
    let _1:x:y:_ = crabulate 1000000 10000000 cups
    print (x * y)
